### ISSUE TITLE
Missing StrengthReduction for complex expressions in actions

### ISSUE THREAD

**Issue by fruffy** [p4lang/p4c#2279]
It looks like `StrengthReduction` is not triggered here:

```p4
control ingress(inout Headers hdr, inout Metadata meta,
                 inout standard_metadata_t stdmeta) {
    simple_struct test = { 128w0 };
    bit<16> tmp_val = 16w3;
    action do_action() {
        hdr.eth_hdr.eth_type = 16w3 + (tmp_val > 2 ? 16w3 : 16w1);
    }
    apply {
        do_action();
    }
}
```
after `MidEndLast` the program looks like this:

```p4 
control ingress(inout Headers hdr, inout Metadata meta, inout standard_metadata_t stdmeta) {
    bit<16> tmp_val_0;
    @name("ingress.do_action") action do_action() {
        hdr.eth_hdr.eth_type = 16w3 + (tmp_val_0 > 16w2 ? 16w3 : 16w1);
    }
    apply {
        tmp_val_0 = 16w3;
        do_action();
    }
}
```
Instead of, for example:
```p4 
control ingress(inout Headers hdr, inout Metadata meta, inout standard_metadata_t stdmeta) {
    bit<16> tmp_val_0;
    @name("ingress.do_action") action do_action() {
        hdr.eth_hdr.eth_type = 16w6;
    }
    apply {
        do_action();
    }
}
```

Example program:

[missed_strength_reduction.p4.txt](https://github.com/p4lang/p4c/files/4409547/missed_strength_reduction.p4.txt)

**Comment by mbudiu-vmw** [comment]
There is no interprocedural constant-propagation in the front-end. I think this is not a real issue; after the action inlining pass (done in the mid-end) this optimization should happen. I will close the issue, but if this hasn't happened by the end of the midend please reopen.

**Comment by fruffy** [comment]
I updated the post, this is also not handled in the mid-end.

**Comment by fruffy** [comment]
@mbudiu-vmw I cannot reopen this issue.

**Comment by fruffy** [comment]
This is fixed with #2932.
